File: 889. Construct Binary Tree from Preorder and Postorder Traversal/main.go
package main

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func constructFromPrePost(preorder []int, postorder []int) *TreeNode {
	n := len(postorder)
	postValToIdx := make(map[int]int)
	for i, n := range postorder {
		postValToIdx[n] = i
	}

	var build func(i1, i2, j1 int) *TreeNode
	build = func(i1, i2, j1 int) *TreeNode {
		if i1 > i2 {
			return nil
		}
		root := &TreeNode{Val: preorder[i1]}
		if i1 != i2 {
			leftVal := preorder[i1+1]
			mid := postValToIdx[leftVal]
			leftSize := mid - j1 + 1
			root.Left = build(i1+1, i1+leftSize, j1)
			root.Right = build(i1+leftSize+1, i2, mid+1)
		}
		return root
	}

	return build(0, n-1, 0)
}

File: 889. Construct Binary Tree from Preorder and Postorder Traversal/main.py
# Definition for a binary tree node.
from typing import List, Optional


class Solution:
    def constructFromPrePost(self, preorder: List[int], postorder: List[int]) -> Optional[TreeNode]:
        N = len(postorder)
        post_num_to_idx = {}
        for i, n in enumerate(postorder):
            post_num_to_idx[n] = i
        
        def build(i1, i2, j1):
            if i1 > i2: return None
            root = TreeNode(preorder[i1])
            if i1 != i2:
                left_val = preorder[i1+1]
                mid = post_num_to_idx[left_val]
                left_size = mid - j1 + 1
                root.left = build(i1 + 1, i1 + left_size, j1)
                root.right = build(i1 + left_size + 1, i2, mid + 1)
            return root
        
        return build(0, N - 1, 0)